// @sudoku/stores/game.js
import { decodeSencode, encodeSudoku } from '@sudoku/sencode';
import { generateSudoku, solveSudoku } from '@sudoku/sudoku';
import { derived, writable } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';
import { createEmptyGrid } from '../../../domain/utils.js';

function createViewModel(game) {
	const sudoku = game.getSudoku();
	const puzzleGrid = sudoku.getPuzzleGrid();
	const grid = sudoku.getGrid();

	return {
		puzzleGrid,
		grid,
		invalidCells: sudoku.getInvalidCells(),
		won: sudoku.isSolved(),
		canUndo: game.canUndo(),
		canRedo: game.canRedo(),
		sencode: encodeSudoku(puzzleGrid),
	};
}

function createGameStore() {
	let game = createGame({ sudoku: createSudoku(createEmptyGrid()) });
	const { subscribe, set } = writable(createViewModel(game));

	function sync() {
		set(createViewModel(game));
	}

	function guess(move) {
		const changed = game.guess(move);
		if (changed) {
			sync();
		}
		return changed;
	}

	function applyHint(position) {
		if (position.x === null || position.y === null) {
			return false;
		}

		const sudoku = game.getSudoku();
		const currentGrid = sudoku.getGrid();
		if (sudoku.isFixedCell(position.y, position.x) || currentGrid[position.y][position.x] !== 0) {
			return false;
		}

		const solution = solveSudoku(sudoku.getPuzzleGrid());
		return guess({
			row: position.y,
			col: position.x,
			value: solution[position.y][position.x],
		});
	}

	function undo() {
		const changed = game.undo();
		if (changed) {
			sync();
		}
		return changed;
	}

	function redo() {
		const changed = game.redo();
		if (changed) {
			sync();
		}
		return changed;
	}

	return {
		subscribe,

		startNew(difficulty) {
			game = createGame({
				sudoku: createSudoku(generateSudoku(difficulty)),
			});
			sync();
		},

		startCustom(sencode) {
			game = createGame({
				sudoku: createSudoku(decodeSencode(sencode)),
			});
			sync();
		},

		guess,
		applyHint,
		undo,
		redo,

		getSencode() {
			return encodeSudoku(game.getSudoku().getPuzzleGrid());
		},

		getDomainGame() {
			return game;
		},
	};
}

export const gameStore = createGameStore();
export const gamePaused = writable(true);
export const gameWon = derived(gameStore, ($gameStore) => $gameStore.won, false);
